src/NHibernate/Hql/Ast/ANTLR/Tree/FromElementType.cs

```diff
@@ -65,32 +72,26 @@ public virtual IType DataType
 		{
 			get
 			{
-				if (_persister == null)
+				if (_persister == null && _queryableCollection != null)
 				{
-					if (_queryableCollection == null)
-					{
-						return null;
-					}
 					return _queryableCollection.Type;
 				}
-				else
-				{
-					return _entityType;
-				}
+
+				return _type;
 			}
 		}
 
 		public IType SelectType
 		{
 			get
 			{
-				if (_entityType == null)
+				if (!(_type is EntityType entityType))
 				{
 					return null;
 				}
 
 				bool shallow = _fromElement.FromClause.Walker.IsShallowQuery;
-				return TypeFactory.ManyToOne(_entityType.GetAssociatedEntityName(), shallow);
+				return TypeFactory.ManyToOne(entityType.GetAssociatedEntityName(), shallow);
```

Comment:
Can we cache it? Maybe calculate it once right inside ctor? It seems it's called quite extensively when query is processed

Reply:
> It seems SelectType is called quite extensively when query is processed

Actually no. I'm wrong about "quite extensively". Still it's executed twice from `InitializeDerivedSelectClause`. Can you refactor it to avoid double object creation? Maybe something like:
```C#
IType type;
if (fromElement.IsFetch || fromElement.IsCollectionOfValuesOrComponents || ((type = fromElement.SelectType) == null))
...
node.DataType = type;
```

Reply:
Suggestion applied.